### PR TITLE
feat(sdk-coin-polygon): support evm based cross chain recovery

### PR DESCRIPTION
**Pull request by DinshawKothari**
Ticket: BG-69445

### CODE REVIEW

File: modules/sdk-coin-polygon/src/polygon.ts [hunk 1]
```diff
@@ -481,6 +488,222 @@ export class Polygon extends Eth {
...
+      // function arguments
+      optionalDeps.ethAbi.rawEncode(types, values),
+    ]);
+  };
+
+  /**
+   * Recovers a tx with non-TSS keys
+   * same expected arguments as recover method
+   */
+  protected async recoverEthLikeforEvmBasedRecovery(
```
Comment (mullapudipruthvik) [inline modules/sdk-coin-polygon/src/polygon.ts:548]: are we missing changes for when funds stuck on eth chain but on polygon wallet?

File: modules/sdk-coin-polygon/test/unit/polygon.ts [hunk 2]
```diff
@@ -763,4 +763,164 @@ describe('Polygon', function () {
...
+      transaction.should.have.property('walletContractAddress');
+      transaction.walletContractAddress.should.equal(TestBitGo.V2.TEST_ETH_WALLET_FIRST_ADDRESS);
+      transaction.should.have.property('recipients');
+      const recipient = transaction.recipients[0];
+      recipient.should.have.property('address');
+      recipient.address.should.equal(destinationAddress);
+      recipient.should.have.property('amount');
+      recipient.amount.should.equal('9999999999999999928');
+      transaction.should.have.property('amount');
+      transaction.amount.should.equal('9999999999999999928');
```
Comment (mullapudipruthvik) [inline modules/sdk-coin-polygon/test/unit/polygon.ts:817]:
```suggestion
  
```

File: modules/sdk-coin-polygon/test/unit/polygon.ts [hunk 3]
```diff
@@ -763,4 +763,164 @@ describe('Polygon', function () {
...
+    });
+
+    it('should generate an unsigned recovery txn for custody wallet', async function () {
```
Comment (mullapudipruthvik) [inline modules/sdk-coin-polygon/test/unit/polygon.ts:820]: What params are different for cold and custody wallet?

File: modules/sdk-coin-polygon/test/unit/polygon.ts [hunk 4]
```diff
@@ -763,4 +763,164 @@ describe('Polygon', function () {
...
+      nock(baseUrl).get('/api').query(mockData.getBalanceRequest(feeAddress)).reply(200, mockData.getBalanceResponse);
+      nock(baseUrl)
+        .get('/api')
+        .query(mockData.getBalanceRequest(walletContractAddress))
+        .reply(200, mockData.getBalanceResponse);
+      nock(baseUrl).get('/api').query(mockData.getContractCallRequest).reply(200, mockData.getContractCallResponse);
+
+      const transaction = (await basecoin.recover({
+        userKey: userXpub,
+        backupKey: '',
```
Comment (mullapudipruthvik) [inline modules/sdk-coin-polygon/test/unit/polygon.ts:789]: Do we need to send empty string ?

File: modules/sdk-coin-polygon/src/polygon.ts [hunk 5]
```diff
@@ -481,6 +488,222 @@ export class Polygon extends Eth {
...
+    params: RecoverOptions
+  ): Promise<RecoveryInfo | OfflineVaultTxInfo> {
+    this.validateEvmBasedRecoveryParams(params);
+
+    // Clean up whitespace from entered values
+    const userKey = params.userKey?.replace(/\s/g, '');
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-polygon/src/polygon.ts:554]:
```suggestion
    const userKey = params.userKey.replace(/\s/g, '');
```
Required field.

File: modules/sdk-coin-eth/src/eth.ts [hunk 6]
```diff
@@ -220,6 +235,8 @@ export interface RecoverOptions {
   eip1559?: EIP1559;
   replayProtectionOptions?: ReplayProtectionOptions;
   isTss?: boolean;
+  feeAddress?: string;
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-eth/src/eth.ts:238]: `feeAddress` cannot be optional as it is used in `getAddressNonce`

Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-eth/src/eth.ts:238]: `feeAddress` cannot be optional as it is used in `getAddressNonce` and `queryAddressBalance`

Comment (mmcshinsky-bitgo) [reply modules/sdk-coin-eth/src/eth.ts:238]: May need to extend this with another type or check if added as required does not create a breaking change.

File: modules/sdk-coin-polygon/src/polygon.ts [hunk 7]
```diff
@@ -481,6 +488,222 @@ export class Polygon extends Eth {
...
+    // build recipients object
+    const recipients: Recipient[] = [
+      {
+        address: params.recoveryDestination,
+        amount: new BigNumber(txAmount).minus(bitgoFeeAmount).toString(10),
+      },
+    ];
+    if (bitgoFeePercentage > 0) {
+      recipients.push({
+        address: bitgoFeeAddress,
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-polygon/src/polygon.ts:613]: Possible bad `undefined` push address since `bitgoFeeAddress` is optional?

**Review by mmcshinsky-bitgo — CHANGES_REQUESTED** [review]

File: modules/sdk-coin-polygon/test/unit/polygon.ts (hunk 3 above)
Comment (DinshawKothari) [reply modules/sdk-coin-polygon/test/unit/polygon.ts:820]: user public key will be provided in cold wallet but not in custody wallet.

File: modules/sdk-coin-eth/src/eth.ts (hunk 6 above)
Comment (DinshawKothari) [reply modules/sdk-coin-eth/src/eth.ts:238]: Making it directly as a required field in RecoverOptions will break other types of recovery.

As there are different types of recovery happening through this recover method I have defined feeAddress as optional.
Particularly for evm cross chain recovery this field is required and hence I have used validateEvmBasedRecoveryParams() in recoverEthLikeforEvmBasedRecovery() which checks that feeAddress is required.

**Review by mmcshinsky-bitgo — DISMISSED** [review]

File: modules/sdk-coin-polygon/src/polygon.ts (hunk 1 above)
Comment (mullapudipruthvik) [reply modules/sdk-coin-polygon/src/polygon.ts:554]: this is not addressed yet

File: modules/statics/src/networks.ts [hunk 8]
```diff
@@ -695,6 +695,7 @@ class Polygon extends Mainnet implements EthereumNetwork {
   forwarderImplementationAddress = '0x5397d0869aba0d55e96d5716d383f6e1d8695ed7';
   walletFactoryAddress = '0xa7198f48c58e91f01317e70cd24c5cce475c1555';
   walletImplementationAddress = '0xe5dcdc13b628c2df813db1080367e929c1507ca0';
+  batcherContractAddress = '0x..'; // TODO: add mainnet contract here
```
Comment (mullapudipruthvik) [inline modules/statics/src/networks.ts:698]: use this address for mainnet
`0x7adc9b3d7521710321bec7dd6897d337e53c2493`

**Review by LeoKotschenreuther — DISMISSED** [review]

File: modules/sdk-coin-eth/src/eth.ts [hunk 9]
```diff
@@ -1204,6 +1234,246 @@ export class Eth extends BaseCoin {
...
+    const txAmount = await this.queryAddressBalance(params.walletContractAddress);
+
+    const bitgoFeePercentage = 0; // TODO: can change the fee% here.
+    const bitgoFeeAmount = txAmount * (bitgoFeePercentage / 100);
+
+    // build recipients object
+    const recipients: Recipient[] = [
+      {
+        address: params.recoveryDestination,
+        amount: new BigNumber(txAmount).minus(bitgoFeeAmount).toString(10),
```
Comment (mullapudipruthvik) [inline modules/sdk-coin-eth/src/eth.ts:1303]: Shouldn't this be `Fixed()`

File: modules/sdk-coin-eth/src/eth.ts [hunk 10]
```diff
@@ -1204,6 +1234,246 @@ export class Eth extends BaseCoin {
...
+
+    const txInfo = {
+      recipients: recipients,
+      expireTime: this.getDefaultExpireTime(),
+      contractSequenceId: sequenceId,
+      gasLimit: gasLimit.toString(10),
+      isEvmBasedCrossChainRecovery: true,
+    };
+
+    const polygonNetwork = this.getNetwork();
```
Comment (mullapudipruthvik) [inline modules/sdk-coin-eth/src/eth.ts:1339]:
```suggestion
    const network = this.getNetwork();
```

**Review by mullapudipruthvik — DISMISSED** [review]
LGTM just have few questions

File: modules/sdk-coin-eth/src/eth.ts [hunk 11]
```diff
@@ -1204,6 +1234,246 @@ export class Eth extends BaseCoin {
...
+        `Fee address ${bitgoFeeAddressBalance} has balance ${(bitgoFeeAddressBalance / weiToGwei).toString()} Gwei.` +
+          `This address must have a balance of at least ${(totalGasNeeded / weiToGwei).toString()}` +
+          ` Gwei to perform recoveries. Try sending some MATIC to this address then retry.`
+      );
+    }
+
+    // get balance of wallet
+    const txAmount = await this.queryAddressBalance(params.walletContractAddress);
+
+    const bitgoFeePercentage = 0; // TODO: can change the fee% here.
```
Comment (mmcshinsky-bitgo) [inline modules/sdk-coin-eth/src/eth.ts:1296]: Can we get a ticket number added to this TODO?

**Review by mullapudipruthvik — DISMISSED** [review]

**Review by mullapudipruthvik — COMMENTED** [review]
Please add support for tokens as well

**Review by LeoKotschenreuther — DISMISSED** [review]

**Comment by DinshawKothari** [comment]
> Please add support for tokens as well

Done. Please review.

**Review by LeoKotschenreuther — APPROVED** [review]

File: modules/sdk-coin-eth/src/eth.ts [hunk 12]
```diff
@@ -1204,6 +1230,369 @@ export class Eth extends BaseCoin {
...
+        gasPrice: optionalDeps.ethUtil.bufferToInt(gasPrice).toFixed(),
+        gasLimit: optionalDeps.ethUtil.bufferToInt(gasLimit).toFixed(),
+      };
+      response['feesUsed'] = feesUsed;
+    }
+
+    return response;
+  }
+
+  async recoverEthLikeTokenforEvmBasedRecovery(
```
Comment (mullapudipruthvik) [inline modules/sdk-coin-eth/src/eth.ts:1421]: I can see a lot of code duplication which can be eliminated
I'm okay if you can do it in a follow up PR

**Review by mullapudipruthvik — APPROVED** [review]

**Review by byhow — APPROVED** [review]

File: modules/sdk-coin-eth/src/eth.ts (hunk 12 above)
Comment (DinshawKothari) [reply modules/sdk-coin-eth/src/eth.ts:1421]: Yeah, I agree. I will do code refactoring in a follow-up PR.

**Pull request merged**